Issue #4383 - avoid NPE from MultiPart Cleanup

lachlan-roberts opened the pull request:
Always assign _parts when constructed so it is never null.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

Closes #4383

lachlan-roberts committed:
Issue #4383 - avoid NPE from MultiPart Cleanup

Always assign _parts when constructed so it is never null.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

janbartel reviewed (approved):
Looks fine.

lachlan-roberts committed:
Issue #4383 - declare some MultiPartFormInputStream fields as final

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Issue #4383 - mark other MultiPartFormInputStream fields as volatile

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
Issue #4383 - synchronize on getParts() and getPart() methods

This will stop two threads from parsing at the same time.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
avoid creating BufferedInputStream if ServletInputStream isFinished

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

janbartel commented on jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartFormInputStream.java:
```diff
@@ -59,18 +60,17 @@
-    private boolean _writeFilesWithFilenames;
-    private boolean _parsed;
-    private int _bufferSize = 16 * 1024;
+    private final MultiMap<Part> _parts = new MultiMap<>();
```
Remind me why this is a multimap and not a HashMap<String, Part>? Can multiple parts have the same name?

lachlan-roberts replied:
Yes, there can be different parts with the same name so we use MultiMap instead of HashMap.

janbartel commented on jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartFormInputStream.java:
```diff
@@ -379,7 +382,7 @@ public boolean isEmpty()
     public Collection<Part> getParsedParts()
     {
-        if (_parts == null)
+        if (_parts.isEmpty())
```
Need synchronized here?

janbartel commented on jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartFormInputStream.java:
```diff
@@ -492,9 +497,6 @@ protected void parse()
         Handler handler = new Handler();
```
To save an unnecessary object creation, only make the Handler if you're going to parse.

lachlan-roberts replied:
This is just below the check where we decide to parse or not, so this is only created once we have decided to parse.

lachlan-roberts committed:
synchronize on deprecated method getParsedParts() as well

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts committed:
delay creation of handler until the MultiPartParser is created

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

janbartel commented on jetty-server/src/main/java/org/eclipse/jetty/server/Request.java:
```diff
@@ -2310,7 +2310,7 @@ public Part getPart(String name) throws IOException, ServletException
     }
 
-    private Collection<Part> getParts(MultiMap<String> params) throws IOException
+    private synchronized Collection<Part> getParts(MultiMap<String> params) throws IOException
```
As per offline discussion, lets not insert synchronization into the request, which is not synchronized for any other method - the spec does not have any support for thread safety in calling methods on requests.

janbartel reviewed (changes requested):
As per offline discussion, lets remove the synchronization and just ensure that creation and cleaning of parts is sane.

lachlan-roberts committed:
Issue #4383 - use atomic state for multipart cleanup

- Removed synchronization for parsing by two threads.

- Introduced an atomic state to decide when it is safe to remove
the parts. The call to deleteParts will now cancel the parsing and
only delete the parts when the parser exits.

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts commented:
@janbartel I added an atomic state to decide when it is safe to remove the parts. 

When you call `deleteParts()` we will move to `ERROR` state which will cause an error in the parser before the next read, the parser will eventually move into `COMPLETED` state and that is when `deleteParts()` knows it is safe to delete the parts.

janbartel commented on jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartFormInputStream.java:
```diff
@@ -58,19 +60,27 @@
+    private volatile File _tmpDir;
+    private volatile boolean _deleteOnExit;
+    private volatile boolean _writeFilesWithFilenames;
+    private volatile boolean _parsed;
```
You don't need _parsed if you have the state variable.

janbartel commented on jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartFormInputStream.java:
```diff
@@ -333,36 +343,36 @@ public String getContentDispositionFilename()
     public boolean isEmpty()
     {
-        if (_parts == null)
+        if (!_parsed)
```
See previous comment.

janbartel commented on jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartFormInputStream.java:
```diff
@@ -397,6 +407,31 @@ public boolean isEmpty()
      */
     public void deleteParts()
     {
+        while (true)
```
Really would prefer not to have these kind of loops.

janbartel commented on jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartFormInputStream.java:
```diff
@@ -397,6 +407,31 @@ public boolean isEmpty()
+                        continue;
+                    break;
+
+                case ERROR:
```
Maybe both ERROR and COMPLETED are both terminal states?

lachlan-roberts replied:
Just moving to `ERROR` state is not enough, the `ERROR` state is indicating that the parsing thread should exit, and the move to `COMPLETED` indicates that it has exited and will not create anymore parts.

janbartel commented on jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartFormInputStream.java:
```diff
@@ -431,15 +466,7 @@ public void deleteParts()
         if (!_parsed)
```
See previous comment

janbartel commented on jetty-http/src/main/java/org/eclipse/jetty/http/MultiPartFormInputStream.java:
```diff
@@ -588,6 +617,27 @@ else if (len == -1)
         }
+        finally
+        {
+            while (true)
```
Again, would really prefer to avoid loops if possible.

lachlan-roberts committed:
parsing thread must call deleteParts if parsing when first deleteParts called
use synchronized state instead of atomic reference

Signed-off-by: Lachlan Roberts <lachlan@webtide.com>

lachlan-roberts commented:
@janbartel nudge

joakime commented:
@lachlan-roberts move these changes over to `jetty-10.0.x` in a new PR.
